### PR TITLE
add in meta sync tool for hudi deltastreamer

### PR DESCRIPTION
**Pull request by the-other-tim-brown**
## What is the purpose of the pull request

#199  Adds a tool for running OneTable sync after each commit in the DeltaStreamer in Hudi

## Brief change log

- Adds sync tool

## Verify this pull request

Added unit test case

### CODE REVIEW

File: hudi-support/extensions/README.md [hunk 1]
```diff
@@ -1,11 +1,22 @@
...
\ No newline at end of file
+3. Run your existing code that use Hudi writers
+
+## Hudi Streamer Extensions
+### When should you use them?
+If you want to use OneTable with Hudi [streaming ingestion](https://hudi.apache.org/docs/hoodie_streaming_ingestion) to sync each commit into other table formats.
+### How do you use them?
+1. Add the extensions jar (`hudi-extensions-0.1.0-SNAPSHOT-bundled.jar`) to your class path.
+2. Add `io.onetable.hudi.sync.OneTableSyncTool` to your list of sync classes
+3. Set the following configurations based on your preferences:
```
Comment (vamshigv) [inline hudi-support/extensions/README.md:20]: same above configs should be set here also right ? can we add a line about it ?

File: hudi-support/extensions/pom.xml [hunk 2]
```diff
@@ -92,6 +106,14 @@
             <groupId>com.esotericsoftware</groupId>
             <artifactId>kryo</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.apache.spark</groupId>
```
Comment (vamshigv) [inline hudi-support/extensions/pom.xml:110]: can & should these be provided ?

File: hudi-support/extensions/README.md (hunk 1 above)
Comment (the-other-tim-brown) [reply hudi-support/extensions/README.md:20]: Above configs are only required for cases where you're writing to an Iceberg target and need to keep the field IDs in the parquet schema

File: hudi-support/extensions/pom.xml (hunk 2 above)
Comment (the-other-tim-brown) [reply hudi-support/extensions/pom.xml:110]: Yes and they are set as provided already at the project level

File: hudi-support/extensions/src/main/java/io/onetable/hudi/sync/OneTableSyncTool.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+            .targetTableFormats(formatsToSync)
+            .hudiSourceConfig(
+                HudiSourceConfig.builder()
+                    .partitionFieldSpecConfig(getPartitionSpecConfig())
+                    .build())
+            .syncMode(SyncMode.INCREMENTAL)
+            .targetMetadataRetentionInHours(
+                config.getInt(OneTableSyncConfig.ONE_TABLE_TARGET_METADATA_RETENTION_HOURS))
+            .build();
+    HudiSourceClientProvider hudiSourceClientProvider = new HudiSourceClientProvider();
```
[the remark this answers is not in the document]
Comment (the-other-tim-brown) [reply hudi-support/extensions/src/main/java/io/onetable/hudi/sync/OneTableSyncTool.java:76]: In practice, this won't make any difference as the syncs are single use and then closed

File: hudi-support/extensions/src/main/java/io/onetable/hudi/sync/OneTableSyncConfig.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+import java.io.Serializable;
+import java.util.Properties;
+
+import org.apache.hudi.common.config.ConfigProperty;
+import org.apache.hudi.sync.common.HoodieSyncConfig;
+
+public class OneTableSyncConfig extends HoodieSyncConfig implements Serializable {
+
+  public static final ConfigProperty<String> ONE_TABLE_FORMATS =
+      ConfigProperty.key("hoodie.onetable.formats")
```
[the remark this answers is not in the document]
Comment (the-other-tim-brown) [reply hudi-support/extensions/src/main/java/io/onetable/hudi/sync/OneTableSyncConfig.java]: updated

File: hudi-support/extensions/README.md (hunk 1 above)
Comment (vamshigv) [reply hudi-support/extensions/README.md:20]: Right but they would still need to add that config as one of the steps here if they want Iceberg as a target right, can we just add that line to be more explicit ?

Comment (the-other-tim-brown) [reply hudi-support/extensions/README.md:20]: It's not related to the hudi sync tool though. It's required on the data writer. I think it is fine as it is.

**Review by vamshigv — APPROVED** [review]

**Pull request merged**